docs/proposals/volumes.md

```diff
@@ -0,0 +1,422 @@
+authors should not have to explicitly account for SELinux except for volume types that require
+special handling of the SELinux context during setup.
+
+SELinux handling for most volumes can be generalized into running a `chcon` operation on the volume
+directory after running the volume plugin's `Setup` function, but there is at least one exception.
+For NFS volumes, the `context` flag must be passed to `mount`, or the `virt_use_nfs` SELinux boolean
+set.  If a system administrator does not wish to set `virt_use_nfs`, the correct context must be
+passed to the `mount` operation in order for the volume to be usable from a container's SELinux
+policy on certain systems.
+
+## Community Design Discussion
+
+- [kubernetes/2630](https://github.com/GoogleCloudPlatform/kubernetes/issues/2630)
+- [kubernetes/11319](https://github.com/GoogleCloudPlatform/kubernetes/issues/11319)
+- [kubernetes/9384](https://github.com/GoogleCloudPlatform/kubernetes/pull/9384)
+
+## Analysis
+
+The system needs to be able to:
+
+1.  Model whether SELinux integration is expected for any pods
+1.  Model correctly which volumes require ownership / SELinux label management
+1.  Determine the correct ownership and SELinux context of each volume in a pod if required
+1.  Set the ownership and permissions on volumes when required
+1.  Relabel volumes with the correct SELinux context when required
+
+### Modeling whether SELinux integration is expected within the cluster
+
+Currently, the SELinux support in kubernetes is essentially optional and is driven by whether
+SELinux is enabled on any particular node.  It should be possible for cluster operators to enforce
+that SELinux is enabled if SELinux integration is expected by the cluster.  If SELinux integration
+is expected and SELinux is not enabled on a node, the Kubelet should not start pods.
+
+There should be a cluster-wide capability that controls whether SELinux integration is enabled in
+the cluster.  Cluster operators generally either run SELinux or they don't; we do not know of any
+cases where an operator wants to have SELinux enforcing on some nodes in a cluster and not on
+others.
+
+There are a couple possibilities for what the behavior of the system should be in this case:
+
+1.  The kubelet could validate that SELinux is enabled at start-up and fail to start if SELinux is
+    expected but not enabled
+2.  The kubelet could create an event for each pod.
+
+### Modeling whether a volume requires ownership/label management
+
+In order to understand what needs to be modeled, we'll need to examine the various facets of
+ownership and SELinux labeling that are required.  For now, let's establish that the types of
+volumes break down into three categories:
+
+1.  the hostPath volume
+2.  volumes created by Kubernetes, which are all derived from empty dir
+3.  distributed filesystems
+
+#### The hostPath corner case:
+
+The hostPath volume should only be used by 'trusted' users, and the permissions of paths exposed
```

Comment:
Very good question, @vishh.  In OpenShift, the SecurityContextConstraint (security policy) handles controlling whether a user can use the `hostPath` volume.  We have a proposal to Kubernetes for this: #7893